File: ios/Kinowo/Models/CinemaFilterSection.swift
import Foundation

/// The Filtry sheet's "Kina" section as pure, view-free data: which boxes read
/// ticked, and what `disabledCinemas` becomes when one is toggled.
///
/// There is ONE cinema filter axis — the `disabledCinemas` exclusion set, shared
/// with the web's `localStorage` (it round-trips through `StateSyncService` and
/// the `?cinema=` deep link). A **flat** city renders `cityCinemas` as a single
/// checkbox list; a **split** city (e.g. London) renders that same universe
/// grouped into `catalog.areas`. Both mutate the one set, mirroring the web's
/// Filtry cinema panel (`buildCinemaPanel` in `shared.js`).
///
/// Pure Foundation → lives in `KinowoCore` and is unit-tested under
/// `swift test`; `FiltersSheet` is a thin renderer over it, and Android's
/// `CinemaFilterSection` is the parallel implementation.
struct CinemaFilterSection {
    /// Tri-state for a checkbox covering one or more cinemas.
    enum Check: Equatable { case on, off, mixed }

    /// The current city's cinema universe + its optional area grouping.
    let catalog: CinemaCatalog
    /// The excluded set, GLOBAL across cities (see `UserPreferences`).
    let disabled: Set<String>

    var cityCinemas: [String] { catalog.cinemas }
    var isSplit: Bool { catalog.isSplit }

    /// How many of THIS city's cinemas are still shown — the count the section
    /// header reads out. Names belonging to other cities are ignored.
    var enabledCount: Int { cityCinemas.filter { !disabled.contains($0) }.count }

    /// The "all cinemas" master checkbox.
    var allCheck: Check { check(of: cityCinemas) }

    func check(ofCinema cinema: String) -> Check {
        disabled.contains(cinema) ? .off : .on
    }

    func check(ofArea area: CinemaArea) -> Check {
        check(of: area.cinemas)
    }

    func check(of cinemas: [String]) -> Check {
        let off = cinemas.reduce(0) { $0 + (disabled.contains($1) ? 1 : 0) }
        if off == 0 { return .on }
        if off == cinemas.count { return .off }
        return .mixed
    }

    /// The new excluded set after ticking/unticking one cinema.
    func setting(cinema: String, enabled: Bool) -> Set<String> {
        setting(cinemas: [cinema], enabled: enabled)
    }

    /// The new excluded set after ticking/unticking a whole area.
    func setting(area: CinemaArea, enabled: Bool) -> Set<String> {
        setting(cinemas: area.cinemas, enabled: enabled)
    }

    /// The new excluded set after the "all cinemas" master. Scoped to this
    /// city's universe so other cities' entries survive a select-all.
    func settingAll(enabled: Bool) -> Set<String> {
        setting(cinemas: cityCinemas, enabled: enabled)
    }

    /// Every mutator funnels here: add/remove exactly the named cinemas and
    /// leave the rest of the (cross-city) set alone.
    private func setting(cinemas: [String], enabled: Bool) -> Set<String> {
        var s = disabled
        if enabled { cinemas.forEach { s.remove($0) } }
        else       { cinemas.forEach { s.insert($0) } }
        return s
    }
}

File: android/app/src/main/java/pl/kinowo/filter/CinemaFilterSection.kt
package pl.kinowo.filter

import pl.kinowo.model.CinemaArea
import pl.kinowo.model.CinemaCatalog

/** Tri-state for a checkbox covering one or more cinemas. */
enum class CinemaCheck { ON, OFF, MIXED }

/**
 * The Filtry sheet's "Kina" section as pure, Compose-free data: which boxes read
 * ticked, and what `disabledCinemas` becomes when one is toggled.
 *
 * There is ONE cinema filter axis — the [disabled] exclusion set, shared with
 * the web's `localStorage` (it round-trips through StateSyncService and the
 * `?cinema=` deep link). A **flat** city renders [cityCinemas] as a single
 * checkbox list; a **split** city (e.g. London) renders that same universe
 * grouped into [CinemaCatalog.areas]. Both mutate the one set, mirroring the
 * web's Filtry cinema panel (`buildCinemaPanel` in `shared.js`).
 *
 * Pure Kotlin → unit-tested by CinemaFilterSectionTest; `FiltersSheet` is a thin
 * renderer over it, and iOS `CinemaFilterSection` is the parallel implementation.
 */
data class CinemaFilterSection(
    /** The current city's cinema universe + its optional area grouping. */
    val catalog: CinemaCatalog,
    /** The excluded set, GLOBAL across cities (see UserPreferences). */
    val disabled: Set<String>,
) {
    val cityCinemas: List<String> get() = catalog.cinemas
    val isSplit: Boolean get() = catalog.isSplit

    /**
     * How many of THIS city's cinemas are still shown — the count the section
     * header reads out. Names belonging to other cities are ignored.
     */
    val enabledCount: Int get() = cityCinemas.count { it !in disabled }

    /** The "all cinemas" master checkbox. */
    val allCheck: CinemaCheck get() = checkOf(cityCinemas)

    fun checkOfCinema(cinema: String): CinemaCheck =
        if (cinema in disabled) CinemaCheck.OFF else CinemaCheck.ON

    fun checkOfArea(area: CinemaArea): CinemaCheck = checkOf(area.cinemas)

    fun checkOf(cinemas: List<String>): CinemaCheck {
        val off = cinemas.count { it in disabled }
        return when (off) {
            0 -> CinemaCheck.ON
            cinemas.size -> CinemaCheck.OFF
            else -> CinemaCheck.MIXED
        }
    }

    /** The new excluded set after ticking/unticking one cinema. */
    fun settingCinema(cinema: String, enabled: Boolean): Set<String> =
        setting(listOf(cinema), enabled)

    /** The new excluded set after ticking/unticking a whole area. */
    fun settingArea(area: CinemaArea, enabled: Boolean): Set<String> =
        setting(area.cinemas, enabled)

    /**
     * The new excluded set after the "all cinemas" master. Scoped to this city's
     * universe so other cities' entries survive a select-all.
     */
    fun settingAll(enabled: Boolean): Set<String> = setting(cityCinemas, enabled)

    /**
     * Every mutator funnels here: add/remove exactly the named cinemas and leave
     * the rest of the (cross-city) set alone.
     */
    private fun setting(cinemas: List<String>, enabled: Boolean): Set<String> =
        if (enabled) disabled - cinemas.toSet() else disabled + cinemas.toSet()
}
